### PR TITLE
Dsp 6106 support vw cli args

### PR DESCRIPTION
**Pull request by jonastim**
Some improvements for the creation of VW models with details in the annotated code

### CODE REVIEW

File: coba/learners/vowpal.py [hunk 1]
```diff
@@ -1,22 +1,22 @@
...
+from coba.primitives import Sparse, Context, Action, Actions
+from coba.utilities import PackageChecker
 
 Feature       = Union[str,int,float]
 Features      = Union[Feature, Sequence[Feature], Dict[str,Union[int,float]]]
 Namespaces    = Dict[str,Features]
 VW_Features   = Sequence[Union[str,int,Tuple[Union[str,int],Union[int,float]],Dict[str,Union[int,float]]]]
 VW_Namespaces = Dict[str,VW_Features]
 
+DEFAULT_NAMESPACE_INTERACTIONS = (1, 'a', 'ax', 'axx')
```
Comment (jonastim) [inline coba/learners/vowpal.py:18]: This wasn't a constant but a (mutable) list default argument which is risky, see for example
```
def dummy(mutable_list = [1,2,3]):
    mutable_list.append(0)
    print(mutable_list)
    
dummy()
[1, 2, 3, 0]
dummy()
[1, 2, 3, 0, 0]
dummy()
[1, 2, 3, 0, 0, 0]
```

File: coba/learners/vowpal.py [hunk 2]
```diff
@@ -196,45 +196,39 @@ class VowpalLearner(Learner):
...
-            kwargs: Any number of additional options to add to the arg string.
+         Args:
+            vw_kwargs: Keyword argument dict that gets translated into VW CLI arguments
+            namespace_interactions: A sequence of namespace interactions to use during learning.
         """
+        args = []
+        vw_kwargs['quiet'] = vw_kwargs.get('quiet', True)
+
+        for k,v in vw_kwargs.items():
+            if v is not False and v is not None:
```
Comment (jonastim) [inline coba/learners/vowpal.py:213]: This condition didn't include `v is not None` but I think for only `"flag_1"` to pass is the more expected behavior
```
{
    "flag_1": True,
    "flag_2": False,
    "flag_3": None
}
```

File: coba/learners/vowpal.py [hunk 3]
```diff
@@ -196,45 +196,39 @@ class VowpalLearner(Learner):
...
 
-        kwargs['quiet'] = kwargs.get('quiet',True)
-
-        for k,v in kwargs.items():
-            if v is not False:
-                k = ("-" if len(k)==1 else "--") + k
-                options.append(k if (v is None or v is True) else f"{k} {v}")
-
-        return " ".join(options)
+        return args
```
Comment (jonastim) [inline coba/learners/vowpal.py:231]: Return args as collection for testability and potential future constructors that don't depend on an assembled config string.

File: coba/learners/vowpal.py [hunk 4]
```diff
@@ -372,25 +366,27 @@ class VowpalEpsilonLearner(VowpalLearner):
...
 
         Args:
             epsilon: The probability that we will explore instead of exploit.
             features: A list of namespaces and interactions to use when learning reward functions.
             seed: The seed used by VW to generate any necessary random numbers.
+            kwargs: Additional key-word args are passed on as VW CLI arguments (unless removed in the function).
         """
-
-        options       = [ "--cb_explore_adf", f"--epsilon {epsilon}" ]
-        noconstant    = sum([f for f in features if isinstance(f,(int,float))]) == 0
```
Comment (jonastim) [inline coba/learners/vowpal.py:388]: Pulled this logic into the `make_args` function

File: coba/tests/test_learners_vowpal.py [hunk 5]
```diff
@@ -64,22 +64,23 @@ def make_examples(self, shared, distincts, labels):
...
         self._finish_calls+=1
 
+
 class VowpalEpsilonLearner_Tests(unittest.TestCase):
 
     @unittest.mock.patch('coba.learners.vowpal.VowpalLearner.__init__')
     def test_defaults(self, mock) -> None:
         VowpalEpsilonLearner()
-        mock.assert_called_once_with("--cb_explore_adf --epsilon 0.05 --interactions ax --interactions axx --ignore_linear x --random_seed 1 --quiet",None)
+        mock.assert_called_once_with("--cb_explore_adf --epsilon 0.05 --random_seed 1 --quiet --interactions ax --interactions axx --ignore_linear x",None)
```
Comment (jonastim) [inline coba/tests/test_learners_vowpal.py:73]: The order changed as kwargs are processed before args are added in `make_args`

**Comment by mrucker** [comment]
So who are you @jonastim? Are you a Microsoft engineer or just someone interested in using coba?

**Comment by mrucker** [comment]
Another great, high-quality code commit. Thank you!

**Comment by jonastim** [comment]
> So who are you @jonastim? Are you a Microsoft engineer or just someone interested in using coba?

He 😄, I am an ML platform engineer at Lyft and we use VW for contextual bandits. I wanted to give our data scientists some better tooling for model development and evaluation. In my research I came across Coba and thought it would be a great platform for that. As we use it I am trying to contribute any improvements back to the project.

**Pull request merged**